### PR TITLE
Set the publicDockerImageReference for Images

### PR DESCRIPTION
**Pull request by mfojtik**
Without having a location of an external registry in the Image object, we can't validate the image signatures for the images when running the command outside the cluster. The verification process has to connect to the registry in order to fetch the manifest and since we only have the internal cluster registry DNS on images, the verification process will fail to connect.

Since we already have the field for ImageStream and since there is no way to know the ImageStream from an image (other than trusting the pull spec), I think it is reasonable to have the field for the Image as well.

Now, the implementation might be a little bit controversial. At the first, the ImageStream use the decorator to set up the DockerImageRepository and PublicDockerImageRepository. However, decorators might not be safe as they mutate the objects and causing a corrupted object in the output (it does not reflect the state in the etcd).

So the options are:

1) Write a controller that will automatically set the fields up for Image/ImageStream, including the DockerImageReference/Repository. Cons: Since the registry location is baked in the master config, this will be one-shot controller so it does not need to run all the time. 

2) Use the PreUpdate and PreCreate in REST to set these fields up for any new or updated image. Cons: This will require a migration for existing images.

3) Use the Decorator() for REST. Cons: Might cause corrupted objects, decorators are generally harmful to controllers, cache, etc..

For the initial implementation, I picked combination of 2 and 3. It means that for existing images, the decorator will set the field up, for new or updated images this will be persisted and stored in etcd.

Not sure which approach is more doable, but I need that field up to unblock the verification flow.
Another option is to make the verification not fetch and verify the manifest (which I believe is already possible), but there might be other use-cases for the public pull reference in image.

### CODE REVIEW

**Comment by mfojtik** [comment]
@smarterclayton @deads2k @soltysh sorry for the lengthy description, but I think this deserves some more thoughts.

File: pkg/image/apis/image/helper.go [hunk 1]
```diff
@@ -1240,3 +1241,59 @@ func ValidateRegistryURL(registryURL string) error {
 	}
 	return nil
 }
+
+// UpdateWithRegistryHostnames takes ImageStreams, ImageStreamList, Image or
+// ImageList object and mutates the image pull specs using configured hostnames
+// for image registry.
+func UpdateWithRegistryHostnames(r RegistryHostnameRetriever, obj runtime.Object) error {
```
Comment (mfojtik) [inline pkg/image/apis/image/helper.go]: @soltysh i added this helper because I hate this code being copied in both IS and image storages and in rest.

**Comment by mfojtik** [comment]
@deads2k one more point for decorators is that changing the registry location requires restart of the API master which will probably cause all external clients/watches to restart/relist, right?

**Comment by smarterclayton** [comment]
A single image can be located from multiple repos, and that could be hundreds of repos (at worst).  Another option 4 would be to create an index for this in a controller and then either use them

If there was a certificate signer controller, it could materialize the index and use those instead.  After all, don't you need signatures from multiple sources?

> Since the registry location is baked in the master config, this will be one-shot controller so it does not need to run all the time.

We have migrations to do this (we have migrate image-references).  This sounds like a variant of that instead of a controller (so option 1a would be 'migrate image-references').

I don't think of images as having an authoritative location.  They just happen to have a preferred location.  Any controller that wants to fetch from a remote system with authentication needs to be able to find multiple locations (because the original user may not have those credentials available).  This really sounds like a controller problem.

**Comment by mfojtik** [comment]
@smarterclayton every image has dockerImageReference set to something like: 

`dockerImageReference: docker-registry.default.svc:5000/foo/bar@sha256:3fb28b3df40a6e1433dda40fad4fdbfac25bd85fcd178257aead0987b7c90cfd`

Everything this PR does is basically add `publicDockerImageReference` field, that has the same value as above but use the external route for registry instead of docker-registry.default...

The reason I need that is so I can run: `oc verify-image-signature sha256:3fb28b3df40a6e1433dda40fad4fdbfac25bd85fcd178257aead0987b7c90cfd` which in turn get the Image object, get the publicDockerImageReference and get the registry part. Then use the registry URL to fetch the manifest for this image.

**Comment by mfojtik** [comment]
after talking to Clayton yesterday, we are not going to do this but instead i'm going to add `--registry-url` flag for `verify-image-signature` to unblock verification via external registry route.

**Pull request closed without merging**